[NOT FOR MERGE] Intel txt aem 2.06 rebased

krystian-hebel opened the pull request:
This PR is made to show and test SKINIT changes on top of what was sent to https://github.com/QubesOS/qubes-grub2/pull/13. After review `intel-txt-aem-2.06-rebased` should be used to overwrite `intel-txt-aem-2.06`, and `-testing` can be removed.

rossphilipson committed:
i386/slaunch: Add basic platform support for secure launch

Some of the commands declared in header files will be implemented in
the follow-up commits.

Signed-off-by: Ross Philipson <ross.philipson@oracle.com>
Signed-off-by: Daniel Kiper <daniel.kiper@oracle.com>
Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

rossphilipson committed:
i386/txt: Add Intel TXT definitions header file

Signed-off-by: Ross Philipson <ross.philipson@oracle.com>
Signed-off-by: Daniel Kiper <daniel.kiper@oracle.com>

rossphilipson committed:
i386/txt: Add Intel TXT core implementation

Signed-off-by: Ross Philipson <ross.philipson@oracle.com>
Signed-off-by: Daniel Kiper <daniel.kiper@oracle.com>
Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>
Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

rossphilipson committed:
i386/txt: Add Intel TXT ACM module support

Signed-off-by: Ross Philipson <ross.philipson@oracle.com>
Signed-off-by: Daniel Kiper <daniel.kiper@oracle.com>
Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

rossphilipson committed:
i386/txt: Add Intel TXT verification routines

Signed-off-by: Ross Philipson <ross.philipson@oracle.com>
Signed-off-by: Daniel Kiper <daniel.kiper@oracle.com>
Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

rossphilipson committed:
i386/slaunch: Add secure launch framework and commands

Signed-off-by: Ross Philipson <ross.philipson@oracle.com>
Signed-off-by: Daniel Kiper <daniel.kiper@oracle.com>
Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

Commit:
i386/memory: Rename PAGE_SHIFT to GRUB_PAGE_SHIFT

...to avoid potential conflicts and confusion.

Signed-off-by: Daniel Kiper <daniel.kiper@oracle.com>

Commit:
i386/memory: Rename PAGE_SIZE to GRUB_PAGE_SIZE and make it global

Subsequent patches will use that constant.

Signed-off-by: Daniel Kiper <daniel.kiper@oracle.com>

Commit:
i386/msr: Merge rdmsr.h and wrmsr.h into msr.h

It does not make sense to have separate headers for separate static
functions. Additionally, we have to add some constants with MSR addresses
in subsequent patches. So, make one common place to store them.

Signed-off-by: Daniel Kiper <daniel.kiper@oracle.com>

Commit:
i386/msr: Rename grub_msr_read() and grub_msr_write()

... to grub_rdmsr() and grub_wrmsr() respectively. New names are more
obvious than older ones.

Signed-off-by: Daniel Kiper <daniel.kiper@oracle.com>

Commit:
i386/tpm: Rename tpm module to tpm_verifier

...to avoid naming collision with TPM TIS and CRB driver introduced
by subsequent patch.

Signed-off-by: Daniel Kiper <daniel.kiper@oracle.com>

Commit:
i386/msr: Extract and improve MSR support detection code

Currently rdmsr and wrmsr commands have own MSR support detection code.
This code is the same. So, it is duplicated. Additionally, this code
cannot be reused by others. Hence, extract this code to a function and
make it public. By the way, improve a code a bit.

Additionally, use GRUB_ERR_BAD_DEVICE instead of GRUB_ERR_BUG to signal
an error because errors encountered by this new routine are not bugs.

Signed-off-by: Daniel Kiper <daniel.kiper@oracle.com>

Commit:
i386/tpm: Add TPM TIS and CRB driver

It will be used by Intel TXT secure launcher introduced
by subsequent patches.

Signed-off-by: Daniel Kiper <daniel.kiper@oracle.com>
Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

Commit:
mmap: Add grub_mmap_get_lowest() and grub_mmap_get_highest()

The functions calculate lowest and highest available RAM
addresses respectively.

Both functions are needed to calculate PMR boundaries for
Intel TXT secure launcher introduced by subsequent patches.

Signed-off-by: Daniel Kiper <daniel.kiper@oracle.com>

krystian-hebel committed:
i386/skinit: Add AMD SKINIT implementation

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

krystian-hebel committed:
i386/slaunch: Add support for AMD SKINIT

This updates `struct grub_slr_entry_dl_info`:
 * by adding `dlme_base` and `dlme_size` which weren't necessary for
   Intel TXT because the same information is passed via OS2SINIT data
 * by changing `dlme_entry` from `grub_uint64_t` to `grub_uint32_t`
   because that should be enough for an offset

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

krystian-hebel committed:
multiboot2: support AMD SKINIT

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

miczyg1 committed:
i386/txt: Initialize TPM 1.2 event log in TXT heap

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>
Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

miczyg1 committed:
multiboot2: Implement TXT slaunch support

The code makes sure that MBI entry goes first in DRTM, so the payload
can measure it first on launch.

SLRT table is allocated on the heap first, size for it is reserved
inside TXT heap by TXT code and data is later copied into its final
place.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>
Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>
Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>
Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

rossphilipson committed:
i386/txt: Initial commit of the Secure Launch Resource Table (SLRT)

Provide definitions of structures and basic functions for constructing
and parsing of SLRT.

Signed-off-by: Ross Philipson <ross.philipson@oracle.com>
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

rossphilipson committed:
i386/txt: switch to using Secure Launch Resource Table interface

Signed-off-by: Ross Philipson <ross.philipson@oracle.com>
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Add GitHub actions for building QubesOS packages

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
.github/workflows/build.yml: run when a tag is pushed

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
.github/workflows/build.yml: drop pull_request_target

This still runs CI twice in a PR.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
multiboot: make GRUB_MULTIBOOT(make_mbi) return MBI's size

GRUB_MULTIBOOT(get_mbi_size) doesn't look like an accurate source of the
final size, more like a minimal memory buffer size.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

krystian-hebel committed:
i386/memory: Define GRUB_PAGE_MASK constant and GRUB_PAGE_{UP,DOWN} macros

Subsequent patches will use those macros and constant.

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

SergiiDmytruk committed:
i386/slaunch: improve error recovery of "slaunch_module" command

Update `slaunch_module` global only if new module passed all checks.

Closes #11.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
i386/txt: do not update MLE header

It's measured and we better measure the binary as is without changing it
on the fly.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
i386/skinit: don't allocate TPM event log low

Do allocation in the same way as it's done for TXT (preferring high
addresses), otherwise Xen loads Dom0 kernel over the TPM even log and
Linux ends up panicking when it detects conflict with e820 map in
Xen-specific code (why wouldn't Xen check it?  who knows).

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

krystian-hebel committed:
.github/workflows/build.yml: bump base-commit

After https://github.com/QubesOS/qubes-grub2/pull/13 got merged, some
of the commits were duplicated, causing build system to fail. Skip
those commits and add the rest starting at the next available
patch-start number.

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel commented:
To see real delta between those branches: https://github.com/TrenchBoot/grub/compare/intel-txt-aem-2.06-testing..intel-txt-aem-2.06-rebased, changes presented by GH in `Files changed` aren't exact because I had to move commits that added CI on top of `base-commit`.

SergiiDmytruk commented on grub-core/loader/i386/skinit.c:
```diff
@@ -0,0 +1,157 @@
+  /* Contrary to the TXT, on AMD we do not have vendor-provided blobs in
+   * reserved memory, we are using normal RAM */
+  err = grub_relocator_alloc_chunk_align (rel, &ch, 0,
+                                          0xffffffff - GRUB_SKINIT_SLB_SIZE,
```
I guess someone during future review could point out that `UP_TO_TOP32()` should be used here as above, could update right away.

SergiiDmytruk commented on grub-core/loader/i386/skinit.c:
This could be squashed into `i386/skinit: Add AMD SKINIT implementation`.

SergiiDmytruk commented on grub-core/loader/i386/skinit.c:
```diff
@@ -0,0 +1,157 @@
+                                          0xffffffff - GRUB_SKINIT_SLB_SIZE,
+                                          GRUB_SKINIT_SLB_SIZE,
+                                          GRUB_SKINIT_SLB_ALIGN,
+                                          GRUB_RELOCATOR_PREFERENCE_LOW, 1);
```
Given that `GRUB_RELOCATOR_PREFERENCE_LOW` caused us problems for TPM log, any benefits of using `GRUB_RELOCATOR_PREFERENCE_HIGH` here as well?  In that case `0x1000000` was used for a base, so maybe there is no danger here assuming availability of aligned 64KiB below 16MiB boundary.

SergiiDmytruk commented on grub-core/loader/i386/skinit.c:
```diff
@@ -0,0 +1,157 @@
+
+  grub_dprintf ("slaunch", "broadcasting INIT\r\n");
+  *apic = 0x000c0500;               // INIT, all excluding self
+  grub_dprintf ("slaunch", "grub_tpm_relinquish_locality\r\n");
```
IIRC this print might be helpful for the delay that it introduces, but the message could probably be edited to say something like "sent INIT broadcasts".

SergiiDmytruk commented on grub-core/loader/i386/slaunch.c:
```diff
@@ -0,0 +1,337 @@
+  if (slp == SLP_NONE)
+    return grub_error (GRUB_ERR_BAD_ARGUMENT, N_("secure launch not enabled"));
+
+  if (slp > SLP_AMD_SKINIT)
```
There should probably be `SLP_END` with the `slp >= SLP_END` check here, because what I've done doesn't look nice.

SergiiDmytruk commented on include/grub/i386/skinit.h:
```diff
@@ -14,22 +14,27 @@
+#define GRUB_SKINIT_SLB_SIZE  0x10000
+#define GRUB_SKINIT_SLB_ALIGN 0x10000
+
+#ifndef ASM_FILE
```
I think this file is never included in an asm-file (directly or indirectly), so this check and defining constants after includes are unnecessary.

SergiiDmytruk reviewed (commented):
Looked at the changes after not working with them for a few weeks and found some things that could be improved while doing the rebase.  Also ran openQA test on Supermicro M11 with the artifacts produced by CI and it passed successfully.

miczyg1 committed:
grub-core/loader/multiboot_elfxx.c: Fix calculating MLE size

To avoid mismatch between OS SINIT data MLE size and MLE size from
MLE header, take the mle_size from MLE header and do not align it
to PMR size, which is 2MB.

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

SergiiDmytruk commented:
[Cleanup](https://github.com/TrenchBoot/trenchboot-issues/issues/48): @krystian-hebel, I think the comments were basically taken care of in `tb-2.12-57-linux-amd` where it's needed for upstreaming, so the branches can be renamed as stated in the OP.

krystian-hebel commented:
> @krystian-hebel, I think the comments were basically taken care of in `tb-2.12-57-linux-amd` where it's needed for upstreaming, so the branches can be renamed as stated in the OP.

Agreed.

krystian-hebel commented:
(had to restore branch to switch target on https://github.com/TrenchBoot/grub/pull/23, removing again)
